projects/interfaces/buffer/buffer.go

```diff
@@ -0,0 +1,41 @@
+package OurBuffer
+
+import "io"
+
+type OurByteBuffer struct {
+	bytes []byte
+}
+
+func NewBuffer(b []byte) *OurByteBuffer {
+	return &OurByteBuffer{bytes: b}
+}
+
+func (b *OurByteBuffer) Bytes() []byte {
+	return b.bytes
+}
```

Comment:
This is a place where documentation is quite important - though the language doesn't have a way of tracking this, there's a concept in programming of _ownership_.

Imagine someone wrote this code:

```go
underlying := []byte("hello")
buf := buffer.NewBuffer(underlying)
underlying[0] = "y"
```

They'd be changing what's inside the buffer from outside of the `OurByteBuffer` type. This can sometimes be desirable (when a type offers a "view" over some data), but often times can be very problematic (e.g. if you have multiple methods on your type which rely on themselves keeping some invariant - this will e.g. be relevant in the "concurrency" project coming up).

Just like [the `bytes.Buffer` type documents](https://pkg.go.dev/bytes#NewBuffer), I would document on `NewBuffer` that `OurByteBuffer` takes ownership of `b` and the caller should never use `b` again after calling this function. Similarly, I'd document that `Bytes()` returns the actual underlying buffer (and maybe mention that the caller should not modify the returned value, or what the consequences would be if they did).

Another approach that can be used is to _copy_ the buffer and return a new value when constructing our type, or when returning from `Bytes()`, but that may be quite expensive to do.

There's a trade-off here between "Doing the thing cheaply, but putting some responsibility on the caller to not cause bugs" vs "Preventing bugs, at the expensive of spending more time and memory copying things".